### ISSUE TITLE
CRAN quantmod_0.4-6 getOptionChain("AAPL")  Error in close.connection(URL) : invalid connection

### ISSUE THREAD

**Issue by AndreMikulec** [joshuaulrich/quantmod#120]
```r
> getOptionChain("AAPL")
Error in close.connection(URL) : invalid connection
```

```r
> Sys.time()
[1] "2016-08-30 06:53:53 CDT"

> getOption("repos")
                              CRAN
"http://mirrors.nics.utk.edu/cran"

> install.packages("quantmod")
Installing package into 'C:/Users/TARGET/R_LIBS_USER_3.3.1'
(as 'lib' is unspecified)
trying URL 'http://mirrors.nics.utk.edu/cran/bin/windows/contrib/3.3/quantmod_0.4-6.zip'
Content type 'application/zip' length 472756 bytes (461 KB)
downloaded 461 KB

package 'quantmod' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
        C:\Users\TARGET\Temp_3.3.1\RtmpEFFHYz\downloaded_packages

> .libPaths()
[1] "C:/Users/TARGET/R_LIBS_USER_3.3.1"
[2] "C:/Users/TARGET/R-Portable_3.3.1/App/R-Portable/library"
>

> library(quantmod)
Loading required package: xts
Loading required package: zoo

Attaching package: 'zoo'

The following objects are masked from 'package:base':

    as.Date, as.Date.numeric

Loading required package: TTR
Version 0.4-0 included new data defaults. See ?getSymbols.
> getOptionChain
function (Symbols, Exp = NULL, src = "yahoo", ...)
{
    Call <- paste("getOptionChain", src, sep = ".")
    if (missing(Exp)) {
        do.call(Call, list(Symbols = Symbols, ...))
    }
    else {
        do.call(Call, list(Symbols = Symbols, Exp = Exp, ...))
    }
}
<environment: namespace:quantmod>
```
The error is here.
```r
> getOptionChain("AAPL")
Error in close.connection(URL) : invalid connection
```
```r
> sessionInfo()
R version 3.3.1 (2016-06-21)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 10586)

locale:
[1] LC_COLLATE=English_United States.1252
[2] LC_CTYPE=English_United States.1252
[3] LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C
[5] LC_TIME=English_United States.1252

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] quantmod_0.4-6 TTR_0.23-1     xts_0.9-7      zoo_1.7-13

loaded via a namespace (and not attached):
[1] tools_3.3.1     grid_3.3.1      jsonlite_1.0    lattice_0.20-33

> search()
 [1] ".GlobalEnv"        "package:quantmod"  "package:TTR"
 [4] "package:xts"       "package:zoo"       "package:stats"
 [7] "package:graphics"  "package:grDevices" "package:utils"
[10] "package:datasets"  "package:methods"   "Autoloads"
[13] "package:base"

>
```

**Comment by AndreMikulec** [comment]
Note: the data 'is there'.

```r
> debug(quantmod:::getOptionChain.yahoo)
> quantmod:::getOptionChain.yahoo("AAPL")
debugging in: quantmod:::getOptionChain.yahoo("AAPL")
```
...
```r

Browse[2]> dftables
$calls
        contractSymbol strike currency lastPrice     change percentChange
1  AAPL160902C00070000   70.0      USD     37.20  0.0000000     0.0000000
2  AAPL160902C00080000   80.0      USD     29.22  0.0000000     0.0000000
...

Browse[2]> n
debug: dftables
Browse[2]> n
Error in close.connection(URL) : invalid connection
...

Browse[2]> args(close.connection)
function (con, type = "rw", ...)
NULL

Browse[2]> urlExp
[1] "https://query2.finance.yahoo.com/v7/finance/options/AAPL"
Browse[2]> URL
                                               description
"https://query2.finance.yahoo.com/v7/finance/options/AAPL"
                                                     class
                                             "url-wininet"
                                                      mode
                                                       "r"
                                                      text
                                                    "text"
                                                    opened
                                                  "closed"
                                                  can read
                                                     "yes"
                                                 can write
                                                      "no"

```

In R/getOptionChain.R 

replacing this

```r
on.exit(close(URL))
```

with this

```r
on.exit(NULL)
```

may? solve the problem?

```r
Browse[2]> close.connection(URL)
Error in close.connection(URL) : invalid connection
Browse[2]> on.exit(NULL)
Browse[2]> n
exiting from: quantmod:::getOptionChain.yahoo("AAPL")
$calls
                    Strike  Last        Chg   Bid   Ask   Vol    OI
AAPL160902C00070000   70.0 37.20  0.0000000 36.60 37.30    10     0
AAPL160902C00080000   80.0 29.22  0.0000000 26.50 27.25     1     5
...
```

**Comment by braverock** [comment]
I believe that this is a duplicate of #92 , which is resolved in current code.

Can you please test the current github code (e.g. via install_github from devtools) and let us know if this resolves your problem?

**Comment by AndreMikulec** [comment]
Seven minutes after braverock response. . . . same result:   Error in close.connection(URL) : invalid connection

Close R.
Open R.

```r
R version 3.3.1 (2016-06-21) -- "Bug in Your Hair"
Copyright (C) 2016 The R Foundation for Statistical Computing
Platform: x86_64-w64-mingw32/x64 (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> Sys.time()
[1] "2016-08-30 07:39:02 CDT"
> devtools:: install_github("joshuaulrich/quantmod")
Downloading GitHub repo joshuaulrich/quantmod@master
from URL https://api.github.com/repos/joshuaulrich/quantmod/zipball/master
Installing quantmod
"C:/Users/TARGET/R-Portable_3.3.1/App/R-Portable/bin/x64/R"  \
  --no-site-file --no-environ --no-save --no-restore --quiet CMD INSTALL  \
  "C:/Users/TARGET/Temp_3.3.1/RtmpUxlP5D/devtools145476e51788/joshuaulrich-quantmod-07885eb"  \
  --library="C:/Users/TARGET/R_LIBS_USER_3.3.1"  \
  --with-keep.source --install-tests

* installing *source* package 'quantmod' ...
** libs

*** arch - i386
C:/Users/TARGET/Rtools34/mingw_32/bin/gfortran      -O3  -mtune=core2 -c ohlcq.f -o ohlcq.o
C:/Users/TARGET/Rtools34/mingw_32/bin/gfortran      -O3  -mtune=core2 -c ohlcz.f -o ohlcz.o
C:/Users/TARGET/Rtools34/mingw_32/bin/gfortran      -O3  -mtune=core2 -c period.max.f -o period.max.o
C:/Users/TARGET/Rtools34/mingw_32/bin/gfortran      -O3  -mtune=core2 -c period.sum.f -o period.sum.o
C:/Users/TARGET/Rtools34/mingw_32/bin/gcc -shared -s -static-libgcc -o quantmod.dll tmp.def ohlcq.o ohlcz.o period.max.o period.sum.o -Ld:/Compiler/gcc-4.9.3/local330/lib/i386 -Ld:/Compiler/gcc-4.9.3/local330/lib -lgfortran -lm -lquadmath -LC:/Users/TARGET/R-PORT~1.1/App/R-PORT~1/bin/i386 -lR
installing to C:/Users/TARGET/R_LIBS_USER_3.3.1/quantmod/libs/i386

*** arch - x64
C:/Users/TARGET/Rtools34/mingw_64/bin/gfortran      -O2  -mtune=core2 -c ohlcq.f -o ohlcq.o
C:/Users/TARGET/Rtools34/mingw_64/bin/gfortran      -O2  -mtune=core2 -c ohlcz.f -o ohlcz.o
C:/Users/TARGET/Rtools34/mingw_64/bin/gfortran      -O2  -mtune=core2 -c period.max.f -o period.max.o
C:/Users/TARGET/Rtools34/mingw_64/bin/gfortran      -O2  -mtune=core2 -c period.sum.f -o period.sum.o
C:/Users/TARGET/Rtools34/mingw_64/bin/gcc -shared -s -static-libgcc -o quantmod.dll tmp.def ohlcq.o ohlcz.o period.max.o period.sum.o -Ld:/Compiler/gcc-4.9.3/local330/lib/x64 -Ld:/Compiler/gcc-4.9.3/local330/lib -lgfortran -lm -lquadmath -LC:/Users/TARGET/R-PORT~1.1/App/R-PORT~1/bin/x64 -lR
installing to C:/Users/TARGET/R_LIBS_USER_3.3.1/quantmod/libs/x64
** R
** demo
** preparing package for lazy loading
Creating a generic function for 'summary' from package 'base' in package 'quantmod'
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded
*** arch - i386
*** arch - x64
* DONE (quantmod)
```
```r
> library(quantmod)
Loading required package: xts
Loading required package: zoo

Attaching package: 'zoo'

The following objects are masked from 'package:base':

    as.Date, as.Date.numeric

Loading required package: TTR
Version 0.4-0 included new data defaults. See ?getSymbols.
```
The error is here.
```r
> getOptionChain("AAPL")
Error in close.connection(URL) : invalid connection
```
```r
>
> sessionInfo()
R version 3.3.1 (2016-06-21)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 10586)

locale:
[1] LC_COLLATE=English_United States.1252
[2] LC_CTYPE=English_United States.1252
[3] LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C
[5] LC_TIME=English_United States.1252

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] quantmod_0.4-6 TTR_0.23-1     xts_0.9-7      zoo_1.7-13

loaded via a namespace (and not attached):
 [1] httr_1.2.1      R6_2.1.2        tools_3.3.1     withr_1.0.2
 [5] curl_0.9.7      memoise_1.0.0   grid_3.3.1      git2r_0.15.0
 [9] jsonlite_1.0    digest_0.6.9    devtools_1.12.0 lattice_0.20-33
```

**Comment by braverock** [comment]
I just had a chance to replicate your error.

Your proposed patch seems to work when applied via fixInNamespace.

I'll do a bit more testing and then apply it permanently.

**Comment by joshuaulrich** [comment]
@braverock The proposed patch to change `on.exit(close(URL))` to `on.exit(NULL)` is not a good solution. It would make more sense to remove the line entirely. However, I am unsure about removing the line entirely because I put that line there to close the `url` connection if the download fails.  Please let me investigate.

**Comment by AndreMikulec** [comment]
In a 'nutshell" in R/getOptionChain.R,  quantmod:::getOptionChain.yahoo,  base::url does not 'open the connection.' This seems 'by design.'  So connection class variable URL is never 'opened.' ( So, it does not have to be 'closed.' )

The isOpen function returns FALSE

```r
library(quantmod)

debug(quantmod:::getOptionChain.yahoo)

quantmod:::getOptionChain.yahoo("AAPL")

...

URL <- url(urlExp)

debug at ... /joshuaulrich-quantmod-07885eb/R/getOptionChain.R#40: URL <- url(urlExp)

Browse[2]> isOpen(URL)
[1] FALSE
```

The case seems that no direct information is stored in R, 
showing whether or not an object of the connection class is open or closed.

```r
Browse[2]> dput(URL)
structure(3L, class = c("url", "connection"), conn_id = <pointer: 0x0000000000000150>)
Browse[2]>
```

It seems that only the Operating System is ask that?

```r
Browse[2]> base::isOpen
function (con, rw = "")
{
    rw <- pmatch(rw, c("read", "write"), 0L)
    .Internal(isOpen(con, rw))
}
<bytecode: 0x000000000ac6bdf8>
<environment: namespace:base>
```

The R variable 'URL' becomes 'txt' becomes 'con'
The following are debugged.

```r
debug(jsonlite::fromJSON)
debug(jsonlite:::fromJSON_string)
debug(jsonlite:::parseJSON)
debug(jsonlite:::parse_con)
```

The following are checked.

```r
is(URL, "connection") # class check
is(txt, "connection") 
is(con, "connection")

isOpen(URL) # Operating System check
isOpen(con)
isOpen(con)
```

To  the 'connnection' class R object ( 'con' in this case ), the jsonlite package author, checks if open, opens, uses,  then 'closes upon function exit.'

```r
Browse[5]> n
debugging in: parse_con(txt, 1024^2, bigint_as_char)
debug: {
    stopifnot(is(con, "connection"))
    if (!isOpen(con)) {                # NEVER OPEN
        open(con, "rb")                # OPEN
        on.exit(close(con))            # CLOSE ON FUNCTION EXIT
    }
    feed_push_parser(readBin(con, raw(), n), reset = TRUE)
    while (length(buf <- readBin(con, raw(), n))) {
        feed_push_parser(buf)
    }
    finalize_push_parser(bigint_as_char)
}
Browse[6]>
```

```r
> sessionInfo()
R version 3.3.1 (2016-06-21)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 7 x64 (build 7601) Service Pack 1

locale:
[1] LC_COLLATE=English_United States.1252
[2] LC_CTYPE=English_United States.1252
[3] LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C
[5] LC_TIME=English_United States.1252

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] quantmod_0.4-6 TTR_0.23-1     xts_0.9-7      zoo_1.7-13

loaded via a namespace (and not attached):
[1] grid_3.3.1      jsonlite_1.0    lattice_0.20-33
> search()
 [1] ".GlobalEnv"        "package:quantmod"  "package:TTR"
 [4] "package:xts"       "package:zoo"       "package:stats"
 [7] "package:graphics"  "package:grDevices" "package:utils"
[10] "package:datasets"  "package:methods"   "Autoloads"
[13] "package:base"
>
```

**Comment by joshuaulrich** [comment]
URLs are only opened when they are actively being read from. The functions that open and read `url` connections are responsible for closing them. But calling `close` on an R connection does not only close the connection; it also destroys the connection.

The problem here is that `close` also destroys the connection, which I did not expect `jsonlite::fromJSON` to do. Which is why I created the following issue: https://github.com/jeroenooms/jsonlite/issues/145. I'm waiting for a response before I decide what to do in `quantmod:::getOptionChain.yahoo`.
